### PR TITLE
Dmadata Migration

### PR DESCRIPTION
**Pull request by hensldm**
Migrates dmadata from an asm to a c file with an included table.

For NON_MATCHING builds, uses the spec to generate the dmadata instead.
This uses mkdmadata from OoT though with 2 modifications.
1. Don't add NOLOAD segments to the dmadata table (doesn't include the buffers and heaps which just recently got their own segments).
2. Segements with the SYMS flag get an UNSET entry instead

Cleans up and adds some docs for DmaMgr.

### CODE REVIEW

**Review by AngheloAlf — APPROVED** [review]

**Comment by Thar0** [comment]
> Segements that end in syms get an UNSET entry instead
This should probably be a spec flag (like `flags NOLOAD`, e.g. `flags SYMS`) instead of using the segment name

File: include/tables/dmadata/dmadata_table_us.h [hunk 1]
```diff
@@ -0,0 +1,1563 @@
+/**
+ * Matching dmadata layout for US
```
Comment (engineer124) [inline include/tables/dmadata/dmadata_table_us.h:2]: Should this specify version number or will the dma table be identical for us-region games?

File: include/z64dma.h [hunk 2]
```diff
@@ -2,16 +2,17 @@
...
 typedef struct {
-    /* 0x00 */ uintptr_t    vromAddr;    // VROM address (source)
-    /* 0x04 */ void*        dramAddr;    // DRAM address (destination)
-    /* 0x08 */ u32          size;        // File Transfer size
-    /* 0x0C */ char*        filename;    // Filename for debugging
-    /* 0x10 */ s32          line;        // Line for debugging
-    /* 0x14 */ s32          unk14;
+    /* 0x00 */ uintptr_t vromAddr; // VROM address (source)
+    /* 0x04 */ void* dramAddr;     // DRAM address (destination)
+    /* 0x08 */ u32 size;           // File Transfer size
```
Comment (engineer124) [inline include/z64dma.h:10]:
```suggestion
    /* 0x08 */ size_t size;           // File Transfer size
```
perhaps?

File: src/dmadata/dmadata.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+// dmadata Table definition
+#define DEFINE_DMA_ENTRY(name, _nameString) \
+    { SEGMENT_ROM_START(name), SEGMENT_ROM_END(name), SEGMENT_ROM_START(name), 0 },
+#define DEFINE_DMA_ENTRY_UNSET(name, _nameString) \
+    { SEGMENT_ROM_START(name), SEGMENT_ROM_END(name), 0xFFFFFFFF, 0xFFFFFFFF },
+
+DmaEntry gDmaDataTable[] = {
+#include "tables/dmadata_table.h"
+    { 0 },
```
Comment (engineer124) [inline src/dmadata/dmadata.c:21]: Is the `{ 0 },` here padding? Could it be made part of `sDmaDataPadding`?

File: include/tables/dmadata/dmadata_table_us.h (hunk 1 above)
Comment (hensldm) [reply include/tables/dmadata/dmadata_table_us.h]: Tbh I'm not 100% sure. I can't imagine them changing file order, but I could be completely wrong. I added 1.0 to the comment header just to clarify.

File: src/dmadata/dmadata.c (hunk 3 above)
Comment (hensldm) [reply src/dmadata/dmadata.c:21]: Nah. Afaik the last entry in the table needs to be all 0's (see `DmaMgr_FindDmaEntry` where it uses `entry->vromEnd == 0` to now it has reached the end)

**Comment by hensldm** [comment]
Just want to note I've added a lot of notes from Tharo's docs over in OoT.

**Review by AngheloAlf — APPROVED** [review]

File: tools/sizes/boot_functions.csv [hunk 4]
```diff
@@ -10,17 +10,17 @@ asm/non_matchings/boot/idle/Idle_ThreadEntry.s,Idle_ThreadEntry,0x80080514,0x33
...
+asm/non_matchings/boot/z_std_dma/DmaMgr_ProcessRequest.s,DmaMgr_ProcessRequest,0x80080A08,0x5F
 asm/non_matchings/boot/z_std_dma/DmaMgr_ThreadEntry.s,DmaMgr_ThreadEntry,0x80080B84,0x20
-asm/non_matchings/boot/z_std_dma/DmaMgr_SendRequestImpl.s,DmaMgr_SendRequestImpl,0x80080C04,0x23
-asm/non_matchings/boot/z_std_dma/DmaMgr_SendRequest0.s,DmaMgr_SendRequest0,0x80080C90,0x1F
-asm/non_matchings/boot/z_std_dma/DmaMgr_Start.s,DmaMgr_Start,0x80080D0C,0x3D
-asm/non_matchings/boot/z_std_dma/DmaMgr_Stop.s,DmaMgr_Stop,0x80080E00,0xC
+asm/non_matchings/boot/z_std_dma/DmaMgr_RequestAsync.s,DmaMgr_RequestAsync,0x80080C04,0x23
+asm/non_matchings/boot/z_std_dma/DmaMgr_RequestSync.s,DmaMgr_RequestSync,0x80080C90,0x1F
+asm/non_matchings/boot/z_std_dma/DmaMgr_Init.s,DmaMgr_Init,0x80080D0C,0x3D
+asm/non_matchings/boot/z_std_dma/DmaMgr_Destory.s,DmaMgr_Destory,0x80080E00,0xC
```
Comment (engineer124) [inline tools/sizes/boot_functions.csv:23]:
```suggestion
asm/non_matchings/boot/z_std_dma/DmaMgr_Stop.s,DmaMgr_Stop,0x80080E00,0xC
```
Don't know what happened here

File: tools/namefixer.py [hunk 5]
```diff
@@ -986,6 +986,9 @@
     "func_8012CA38": "Gfx_SetupDL59_Opa",
 
     "Game_SetFramerateDivisor": "GameState_SetFramerateDivisor",
+    
+    "DmaMgr_SendRequestImpl": "DmaMgr_SendRequest",
```
Comment (engineer124) [inline tools/namefixer.py:990]:
```suggestion
    "DmaMgr_SendRequestImpl": "DmaMgr_RequestAsync",
```

**Review by engineer124 — APPROVED** [review]

**Review by tom-overton — APPROVED** [review]
You may want to take this approval with a grain of salt, because I'm a little shaky on the part of the build process where the dmadata table is constructed, but the code changes look good to me.

**Pull request merged**